pkg/ds/daemon_set.go

```diff
@@ -0,0 +1,353 @@
+type kpStore interface {
+	SetPod(
+		podPrefix kp.PodPrefix,
+		nodeName types.NodeName,
+		manifest pods.Manifest,
+	) (time.Duration, error)
+
+	DeletePod(podPrefix kp.PodPrefix,
+		nodeName types.NodeName,
+		manifestID types.PodID,
+	) (time.Duration, error)
+}
+
+type daemonSet struct {
+	fields.DaemonSet
+
+	logger         logging.Logger
+	kpStore        kpStore
+	scheduler      scheduler.Scheduler
+	dsStore        dsstore.Store
+	nodeApplicator labels.Applicator
+}
+
+func New(
+	fields fields.DaemonSet,
+	dsStore dsstore.Store,
+	kpStore kpStore,
+	logger logging.Logger,
+	nodeApplicator labels.Applicator,
+	scheduler scheduler.Scheduler,
+) DaemonSet {
+	// TODO: Assert that arguments are not nil
+
+	return &daemonSet{
+		DaemonSet: fields,
+
+		dsStore:        dsStore,
+		kpStore:        kpStore,
+		logger:         logger,
+		nodeApplicator: nodeApplicator,
+		scheduler:      scheduler,
+	}
+}
+
+func (ds *daemonSet) ID() fields.ID {
+	return ds.DaemonSet.ID
+}
+
+// WatchDesires watches for changes to its daemon set, then schedule/unschedule
+// pods to to the nodes that it is responsible for
+//
+// Whatever calls WatchDesires is responsible for sending signals for whether
+// the daemon set updated or deleted
+//
+// When this is first called, it assumes that the daemon set is created
+//
+// The caller is responsible for sending signals when something has been changed
+func (ds *daemonSet) WatchDesires(
+	quitCh <-chan struct{},
+	updatedCh <-chan fields.DaemonSet,
+	deletedCh <-chan struct{},
+	nodesChangedCh <-chan struct{},
+) <-chan error {
+	errCh := make(chan error)
+
+	// Do something whenever something is changed
+	go func() {
+		var err error
+		defer close(errCh)
+
+		// Try to schedule pods when this begins watching
+		if !ds.Disabled {
+			ds.logger.NoFields().Infof("Received new daemon set: %v", *ds)
+			err = ds.addPods()
+			if err != nil {
+				err = util.Errorf("Unable to add pods to intent tree: %v", err)
+			}
+		}
+
+		for {
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-quitCh:
+				}
+			}
+
+			select {
+			case newDS := <-updatedCh:
+				ds.logger.NoFields().Infof("Received daemon set update signal: %v", newDS)
+				if ds.ID() != newDS.ID {
+					err = util.Errorf("Expected uuid to be the same, expected '%v', got '%v'", ds.ID(), newDS.ID)
+					continue
+				}
+				ds.DaemonSet = newDS
+
+				if ds.Disabled {
+					err = ds.clearPods()
+					if err != nil {
+						err = util.Errorf("Unable to clear pods from intent tree: %v", err)
+					}
+					continue
+				}
+				err := ds.removePods()
+				if err != nil {
+					err = util.Errorf("Unable to remove pods from intent tree: %v", err)
+					continue
+				}
+				err = ds.addPods()
+				if err != nil {
+					err = util.Errorf("Unable to add pods to intent tree: %v", err)
+					continue
+				}
+
+			case <-deletedCh:
+				ds.logger.NoFields().Infof("Received daemon set delete signal")
+				err = ds.clearPods()
```

Comment:
The deleted signal is sent when the daemon set is deleted in the store, so we shouldn't need to worry about what its fields were. Which is why we should unschedule it.